sky/provision/gcp/config.py

```diff
@@ -729,12 +748,58 @@ def get_gpu_direct_usable_vpcs_and_subnets(
                                     compute,
                                     network=vpc_name)
         # Apply firewall rules
-        _create_rules(project_id, compute, constants.FIREWALL_RULES_TEMPLATE,
-                      vpc_name)
+        # No need to recreate the rules if exist,
+        # as they are totally managed by SkyPilot,
+        # in this case, we can skip the rules creation during failover
+        _create_rules(project_id,
+                      compute,
+                      constants.FIREWALL_RULES_TEMPLATE,
+                      vpc_name,
+                      recreate=False)
         vpc_subnet_pairs.append((vpc_name, subnets[0]))
     return vpc_subnet_pairs
 
 
+def delete_gpu_direct_vpcs_and_subnets(
+    cluster_name: str,
+    project_id: str,
+    region: str,
+    failover: bool = False,
+):
+    """Delete GPU Direct subnets, firewalls, and VPCs."""
+    compute = _create_compute()
+    cluster_prefix = cluster_name
+
+    # TODO(hailong): Determine the num_vpcs per different GPU Direct types
+    num_vpcs = constants.SKYPILOT_GPU_DIRECT_VPC_NUM
+
+    for i in range(num_vpcs):
+        if i == 0:
+            vpc_name = f'{cluster_prefix}-mgmt-net'
+        else:
+            vpc_name = f'{cluster_prefix}-data-net-{i}'
```

Comment:
Can we move this logic into a shared function? I noticed it's used in multiple places

Reply:
Updated